File: seguridad/seguridad.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>

using namespace std;

vector<int> dijkstra(const unordered_map<int, vector<pair<int, int>>>& graph, int start, int n) {
    vector<int> distances(n, INT_MAX);
    distances[start] = 0;

    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push({0, start});

    while (!pq.empty()) {
        int current_distance = pq.top().first;
        int current_node = pq.top().second;
        pq.pop();

        if (current_distance > distances[current_node]) continue;

        for (const auto& neighbor : graph.at(current_node)) {
            int next_node = neighbor.first;
            int weight = neighbor.second;
            int new_distance = current_distance + weight;

            if (new_distance < distances[next_node]) {
                distances[next_node] = new_distance;
                pq.push({new_distance, next_node});
            }
        }
    }

    return distances;
}

int main() {
    int n, m;
    cin >> n >> m;

    unordered_map<int, vector<pair<int, int>>> graph;

    for (int i = 0; i < m; i++) {
        int v, w, c;
        cin >> v >> w >> c;
        graph[v].push_back({w, c});
        graph[w].push_back({v, c});
    }

    vector<int> dist_start = dijkstra(graph, 0, n);
    vector<int> dist_end = dijkstra(graph, n - 1, n);

    int min_distance = dist_start[n - 1];
    int result = 0;

    for (int i = 0; i < n; i++) {
        for (const auto& edge : graph[i]) {
            int w = edge.first;
            int c = edge.second;
            if (dist_start[i] + c + dist_end[w] == min_distance) {
                result += c * 2;
            }
        }
    }

    cout << result << endl;
    return 0;
}

File: seguridad/seguridad.py
from collections import defaultdict
import heapq

def dijkstra(graph, start, n):
    distances = {node: float('inf') for node in range(n)}
    distances[start] = 0
    
    priority_queue = [(0, start)]
    
    while priority_queue:
        current_distance, current_node = heapq.heappop(priority_queue)
        
        if current_distance > distances[current_node]:
            continue
        
        for neighbor, weight in graph[current_node]:
            distance = current_distance + weight
            
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heapq.heappush(priority_queue, (distance, neighbor))
    
    return distances

def solve():
    n, m = list(map(int, input().split()))

    graph = defaultdict(list)
    
    for _ in range(m):
        v, w, c = list(map(int, input().split()))
        graph[v].append((w, c))
        graph[w].append((v, c))

    dist_from_start = dijkstra(graph, 0, n)
    
    dist_from_end = dijkstra(graph, n-1, n)

    min_distance = dist_from_start[n-1]
    
    total_cost = 0
    for v in range(n):
        for w, c in graph[v]:
            if dist_from_start[v] + c + dist_from_end[w] == min_distance or dist_from_start[w] + c + dist_from_end[v] == min_distance:
                total_cost += c
    
    print(total_cost)

solve()